ingestion-edge/ingestion_edge/publish.py

```diff
@@ -44,21 +43,25 @@
 def _publish(topic: str, data: bytes, attrs: Dict[str, str]) -> str:
     """Publish one message to the topic.
 
-    Uses PublisherClient to try to publish the message in a batch and handles
-    non-transient batch failure by retrying with a single message.
+    Use PublisherClient to publish the message in a batch.
 
-    Note that this method will block until it gets a result, but gunicorn can
-    be configured with worker processes that each use coroutines to handle many
-    requests at once.
+    PublishError from PublisherClient requires special handling because it is a
+    non-transient error. It is raised for the whole batch and does not indicate
+    what message failed or whether all messages failed. Circumvent batching to
+    determine if this message in particular failed.
+
+    Note that this method will block until it gets a result, but gevent workers
+    should yield the coroutine during IO to handle multiple connections per
+    worker.
     """
     try:
         return client.publish(topic, data, **attrs).result()
     except PublishError:
-        # Batch failed because pubsub rejected at least one message
+        # Batch failed because pubsub permanently rejected at least one message
```

Comment:
So if this is the message that was permanently rejected, and this method raises `PublishError`, then that would be caught by the general `except:` clause on line 97, which would cause the message to be `nack`ed. If I'm understanding right, that would mean the message would be put back in the queue and attempted again later? If it is a permanent failure, is that the desired outcome?

Reply:
a message with a permanent failure shouldn't ever end up in the queue, but if it does then it would be retried indefinitely.

The expected behavior is that a message that raises `PublishError` never goes into the queue in the first place, and results in a 500 error at line 134.

Reply:
Could a transient error "outrank" a `PublishError`? That is, could you get a message that at first causes a transient error to happen, because of some outage, and gets added to the queue, and then in `flush` the message gets processed again, and causes a permanent failure? If that happened, the queue would fill up with permanent-error-causing messages.

Reply:
hmm, you're right, that is definitely broken. I will file an issue to fix that bug and add a test case for it.

Reply:
filed #112